cylc/flow/platforms.py

```diff
@@ -16,65 +16,64 @@
 #
 # Tests for the platform lookup.
 
+from copy import deepcopy
 import re
+
 from cylc.flow.exceptions import PlatformLookupError
+from cylc.flow.cfgspec.glbl_cfg import glbl_cfg
 
 
-def forward_lookup(platforms, job_platform):
+def platform_from_name(platform_name=None, platforms=None):
     """
     Find out which job platform to use given a list of possible platforms and
     a task platform string.
 
     Verifies selected platform is present in global.rc file and returns it,
-    raises error if platfrom is not in global.rc or returns 'localhost' if
+    raises error if platform is not in global.rc or returns 'localhost' if
     no platform is initally selected.
 
     Args:
-        job_platform (str):
-            platform item from config [runtime][TASK]platform
-        platforms (dictionary):
-            list of possible platforms defined by global.rc
-
+        platform_name (str):
+            name of platform to be retrieved.
+        platforms ():
+            globalrc platforms given as a dict for logic testing purposes
     Returns:
-        platform (str):
-            string representing a platform from the global config.
-
-    Example:
-    >>> platforms = {
-    ...     'suite server platform': None,
-    ...     'desktop[0-9][0-9]|laptop[0-9][0-9]': None,
-    ...     'sugar': {
-    ...         'remote hosts': 'localhost',
-    ...         'batch system': 'slurm'
-    ...     },
-    ...     'hpc': {
-    ...         'remote hosts': ['hpc1', 'hpc2'],
-    ...         'batch system': 'pbs'
-    ...     },
-    ...     'hpc1-bg': {
-    ...         'remote hosts': 'hpc1',
-    ...         'batch system': 'background'
-    ...     },
-    ...     'hpc2-bg': {
-    ...         'remote hosts': 'hpc2',
-    ...         'batch system': 'background'
-    ...     }
-    ... }
-    >>> job_platform = 'desktop22'
-    >>> forward_lookup(platforms, job_platform)
-    'desktop22'
+        platform (dict):
+            object containing settings for a platform, loaded from
+            Global Config.
+
+    TODO:
+        - Refactor testing for this method. Ideally including a method
+          example.
+        - Work out what to do with cases where localhost not set.
     """
-    if job_platform is None:
-        return 'localhost'
-    for platform in reversed(list(platforms)):
-        if re.fullmatch(platform, job_platform):
-            return job_platform
+    if platforms is None:
+        platforms = glbl_cfg().get(['platforms'])
+
+    if platform_name is None:
+        platform_data = deepcopy(platforms['localhost'])
+        platform_data['name'] = 'localhost'
```

Comment:
It would be good to add the `name` field when the config is read in to avoid the need mutate the platform dict (and hence the need to call `deepcopy`).

If time consuming, feel free to punt it to a follow-up issue.

Reply:
https://github.com/wxtim/cylc/pull/7

Reply:
I think that it needs to be left for now. It turned out to be non-trivial.